File: shesha-reactjs/src/interfaces/dataTypes.ts
export const ArrayFormats = {
  /** Simple values (stored as Json) */
  simple: 'simple',
  /** Referencing entities */
  entityReference: 'entity',
  /** Many to many Entities */
  manyToManyEntities: 'many-entity',
  /** Child Entities (regular Entity but used only as part of parent Entity) */
  childEntities: 'child-entity',
  /** Child objects (any Json object) */
  childObjects: 'object',
  /**  Multi value Reference list item */
  multivalueReferenceList: 'multivalue-reference-list',
};

export const ObjectFormats = {
  object: 'object',
  interface: 'interface',
};

export const StringFormats = {
  singleline: 'singleline',
  multiline: 'multiline',
  html: 'html',
  json: 'json',
  javascript: 'javascript',
  password: 'password',
  emailAddress: 'email',
  phoneNumber: 'phone',
  url: 'url',
};

export const NumberFormats = {
  float: 'float',
  double: 'double',
  int32: 'int32',
  int64: 'int64',
  decimal: 'decimal',
};

export const DataTypes = {
  guid: 'guid',
  string: 'string',
  number: 'number',
  date: 'date',
  time: 'time',
  dateTime: 'date-time',
  entityReference: 'entity',
  file: 'file',
  referenceListItem: 'reference-list-item',
  boolean: 'boolean',
  array: 'array',
  object: 'object',
  geometry: 'geometry',
  advanced: 'advanced',
  specification: 'specification',
  context: 'context',
  function: 'function',
  any: 'any',
  AllowedComponents: (dataType: string, dataFormat: string): string[] => {
    switch (dataType) {
      case DataTypes.string:
        if (dataFormat === StringFormats.multiline || dataFormat === StringFormats.html || dataFormat === StringFormats.javascript || dataFormat === StringFormats.json)
          return ['textArea', 'text'];
        return ['textField', 'text', 'dropdown'];
      case DataTypes.number:
        return ['numberField', 'slider', 'text', 'textField'];
      case DataTypes.date:
      case DataTypes.dateTime:
        return ['dateField', 'text', 'textField'];
      case DataTypes.time:
        return ['timePicker', 'text', 'textField'];
      case DataTypes.boolean:
        return ['switch', 'checkbox', 'text', 'textField'];
      case DataTypes.entityReference:
        return ['autocomplete', 'entityPicker', 'entityReference'];
      case DataTypes.referenceListItem:
        return ['radio', 'dropdown'];
      case DataTypes.file:
        return ['fileUpload'];
      case DataTypes.object:
        return ['subForm'];
      case DataTypes.advanced:
        return [dataFormat];
      case DataTypes.array:
        switch (dataFormat) {
          case ArrayFormats.entityReference:
            return ['autocomplete', 'entityPicker'];
          case ArrayFormats.manyToManyEntities:
            return ['autocomplete', 'entityPicker'];
          case ArrayFormats.childObjects:
            return ['childEntitiesTagGroup'];
          case ArrayFormats.multivalueReferenceList:
            return ['dropdown'];
        }
    }
    return [];
  },
  allowedCompoenentsCode: `
  const allowedComponents = (dataType, dataFormat) => {
    switch (dataType) {
      case 'string': 
        if (dataFormat === 'multiline' || dataFormat === 'html' || dataFormat === 'javascript' || dataFormat === 'json')
          return ['textArea', 'text'];
        return ['textField', 'text', 'dropdown'];
      case 'number':
          return ['numberField', 'slider', 'text', 'textField'];
      case 'date':
      case 'date-time':
        return ['dateField', 'text', 'textField'];
      case 'time':
        return ['timePicker', 'text', 'textField'];
      case 'boolean':
        return ['switch', 'checkbox', 'text', 'textField'];
      case 'entity':
        return ['autocomplete', 'entityPicker', 'entityReference'];
      case 'reference-list-item':
        return ['radio', 'dropdown'];
      case 'file':
        return ['fileUpload'];
      case 'object':
        return ['subForm'];
      case 'advanced':
          return [dataFormat];
      case 'array':
        switch (dataFormat) {
          case 'entity':
            return ['autocomplete','entityPicker'];
          case 'many-entity':
            return ['autocomplete','entityPicker'];
          case 'object':
            return ['childEntitiesTagGroup'];
          case 'multivalue-reference-list':
            return ['dropdown'];
        }
    }
    return [];
  }`,
};

File: shesha-core/src/Shesha.Framework/Metadata/DataTypes.cs
using System;

namespace Shesha.Metadata
{
    /// <summary>
    /// Data types
    /// </summary>
    public static class DataTypes
    {
        public const string Guid = "guid";
        public const string String = "string";
        public const string Number = "number";
        public const string Date = "date";
        public const string Time = "time";
        public const string DateTime = "date-time";
        public const string EntityReference = "entity";
        public const string File = "file";
        public const string ReferenceListItem = "reference-list-item";
        public const string Boolean = "boolean";
        public const string Array = "array";
        public const string Object = "object";
        public const string Geometry = "geometry";

        // DataFormat for this data type should contain the name of the UI component (it is used only for binding to the UI and is not used in the backend)
        public const string Advanced = "advanced";
    }


    /// <summary>
    /// Custom array formats
    /// </summary>
    public static class ArrayFormats
    {
        /// <summary>
        /// Simple values (stored as Json)
        /// </summary>
        public const string Simple = "simple";
        /// <summary>
        /// Referencing entities
        /// </summary>
        public const string EntityReference = "entity";
        /// <summary>
        /// Many to many Entities
        /// </summary>
        public const string ManyToManyEntities = "many-entity";
        /// <summary>
        /// Child Entities (regular Entity but used only as part of parent Entity)
        /// </summary>
        public const string ChildEntities = "child-entity";
        /// <summary>
        /// Child objects (any Json object)
        /// </summary>
        public const string ChildObjects = "object";
        /// <summary>
        /// Multi value Reference list item
        /// </summary>
        public const string MultivalueReferenceList = "multivalue-reference-list";
    }

    /// <summary>
    /// Custom string formats
    /// </summary>
    public static class StringFormats 
    {
        public const string Singleline = "singleline";
        public const string Multiline = "multiline";
        public const string Html = "html";
        public const string Json = "json";
        public const string Javascript = "javascript";
        public const string Password = "password";
        public const string EmailAddress = "email";
        public const string PhoneNumber = "phone";
        public const string Url = "url";        
    }

    /// <summary>
    /// Custom number formats
    /// </summary>
    public static class NumberFormats
    {
        public const string Float = "float";
        [Obsolete("Use 'float' instead")]
        public const string Double = "double";
        public const string Int32 = "int32";
        public const string Int64 = "int64";
        public const string Decimal = "decimal";
    }

    public static class ObjectFormats
    {
        public const string Object = "object";
        public const string Interface = "interface";
    }

    public static class EntityFormats
    {
        public const string Entity = "entity";
        public const string GenericEntity = "generic-entity";
    }
}